src/utils/download.py

```diff
@@ -0,0 +1,134 @@
+#!/usr/bin/python3
+
+import argparse
+import time
+import requests
+import os
+
+
+class OutputSettings:
+    def __init__(self):
+        self.print_hash = True
+        self.print_matches = False
+        self.print_filename = False
+        self.save_to_directory = None
+
+
+def query(mquery_server: str, yara_rule: str) -> str:
+    """ Queries mquery server and returns a new job ID """
+    res = requests.post(
+        f"{mquery_server}/api/query",
+        json={
+            "method": "query",
+            "raw_yara": yara_rule,
+            "taint": None,
+            "priority": "normal",
+            "method": "query",
+        },
+    ).json()
+    if "error" in res:
+        raise RuntimeError("Query error: " + res["error"])
+
+    return res["query_hash"]
+
+
+def process_job(
+    mquery_server: str, job_id: str, output: OutputSettings
+) -> None:
+    offset = 0
+
+    while True:
+        out = requests.get(
+            f"{mquery_server}/api/matches/{job_id}?offset={offset}&limit=50"
+        ).json()
+
+        if "error" in out:
+            raise RuntimeError(out["error"])
+
+        for match in out["matches"]:
+            sha256 = match["meta"]["sha256"]["display_text"]
+            line = sha256
+
+            file_path = match["file"]
+            if output.print_filename:
+                line += f" {file_path}"
+
+            if output.print_matches:
+                for matched_rule in match["matches"]:
+                    line += f" {matched_rule}"
+
+            print(line)
+
+            if output.save_to_directory is not None:
+                with open(
+                    f"{output.save_to_directory}/{sha256}", "wb"
+                ) as outf:
+                    r = requests.get(
+                        f"{mquery_server}/api/download",
+                        {
+                            "job_id": job_id,
+                            "ordinal": offset,
+                            "file_path": file_path,
+                        },
+                    )
+                    outf.write(r.content)
+
+            offset += 1
+
+        if out["job"]["status"] in ["cancelled", "failed", "done", "removed"]:
+            break
+
+        time.sleep(1.0)
```

Comment:
Why do we sleep one second?
Think scaling, for tens of thousands of files, even 1million, it is so much time wasted sleeping :baby: :baby_bottle: :zzz:

Reply:
Right, result of me merging functions into one without thinking. It's because when you create a new job, you usually have to wait for results anyway (they are downloaded when mquery is still processing them).

This code should check if it downloaded 50 (MAX_FILES) files, and if yes than skip sleeping. I'll fix it in a second.

Also, if we wanted to go really fast, we could parallelise file downloads. It's relatively easy with process pools i guess. I'll leave it as a future improvement though.

Reply:
parallel sounds good. Would you open an issue?